File: samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/ops/NodeLivenessObserver.java
package systems.zlink.samples.zoneworld.server.ops;

import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;

import systems.zlink.framework.monitoring.ZLinkMeshNodeSnapshot;
import systems.zlink.framework.monitoring.ZLinkMeshPeerSnapshot;
import systems.zlink.framework.monitoring.ZLinkObservedStatus;
import systems.zlink.framework.monitoring.ZLinkPeerState;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;
import systems.zlink.samples.zoneworld.server.configuration.NodeRegistry;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.Flow;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Whether a zone node is registered and its link is up. Neither is a question the console can ask:
 * a node that has stopped is not there to answer, so the console learns it from the mesh runtime it
 * already observes in its own process.
 */
public final class NodeLivenessObserver implements ApplicationRunner {
    private final ZLinkRouteMeshRuntime runtime;
    private final NodeRegistry registry;
    // The runtime signals its observation publisher through a weak reference, so the
    // stream stops as soon as the publisher is collected. The console watches node
    // lifecycles for as long as it runs, so it keeps the publisher for that long.
    private Flow.Publisher<ZLinkObservedStatus<ZLinkMeshNodeSnapshot>> observation;
    private final ScheduledExecutorService expiry =
            Executors.newSingleThreadScheduledExecutor(
                    runnable -> {
                        Thread thread = new Thread(runnable, "zoneworld-report-expiry");
                        thread.setDaemon(true);
                        return thread;
                    });

    public NodeLivenessObserver(ZLinkRouteMeshRuntime runtime, NodeRegistry registry) {
        this.runtime = runtime;
        this.registry = registry;
    }

    @Override
    public void run(ApplicationArguments args) {
        // --8<-- [start:doc-zw-snapshot-peers]
        ZLinkMeshNodeSnapshot initial = runtime.snapshot(ZoneWorldNames.MESH);
        registry.applyLiveRoutingIds(readyRoutingIds(initial));
        // --8<-- [end:doc-zw-snapshot-peers]

        // --8<-- [start:doc-zw-observe-peers]
        observation = runtime.observe(ZoneWorldNames.MESH, 32);
        observation.subscribe(
                new Flow.Subscriber<ZLinkObservedStatus<ZLinkMeshNodeSnapshot>>() {
                    @Override
                    public void onSubscribe(Flow.Subscription subscription) {
                        subscription.request(Long.MAX_VALUE);
                    }

                    @Override
                    public void onNext(ZLinkObservedStatus<ZLinkMeshNodeSnapshot> observed) {
                        registry.applyLiveRoutingIds(readyRoutingIds(observed.status()));
                    }

                    @Override
                    public void onError(Throwable error) {
                        System.out.println(
                                "node observation error mesh="
                                        + ZoneWorldNames.MESH
                                        + " detail="
                                        + error.getMessage());
                    }

                    @Override
                    public void onComplete() {}
                });
        // --8<-- [end:doc-zw-observe-peers]
        expiry.scheduleAtFixedRate(registry::expireStaleReports, 1, 1, TimeUnit.SECONDS);
    }

    private static Map<String, Boolean> readyRoutingIds(ZLinkMeshNodeSnapshot status) {
        Map<String, Boolean> observed = new LinkedHashMap<>();
        for (ZLinkMeshPeerSnapshot peer : status.peers()) {
            if (peer.state() != ZLinkPeerState.READY) continue;
            String routingId = peer.nodeRid().toString();
            observed.put(routingId, true);
        }
        return observed;
    }
}

File: samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/server/ops/NodeLivenessObserver.kt
package systems.zlink.samples.kotlin.zoneworld.server.ops

import java.util.concurrent.Executors
import java.util.concurrent.Flow
import java.util.concurrent.TimeUnit
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import systems.zlink.framework.monitoring.ZLinkMeshNodeSnapshot
import systems.zlink.framework.monitoring.ZLinkObservedStatus
import systems.zlink.framework.monitoring.ZLinkPeerState
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime
import systems.zlink.samples.kotlin.zoneworld.server.configuration.NodeRegistry
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

/**
 * Whether a zone node is registered and its link is up. Neither is a question the console can ask:
 * a node that has stopped is not there to answer, so the console learns it from the mesh runtime it
 * already observes in its own process.
 */
class NodeLivenessObserver(
    private val runtime: ZLinkRouteMeshRuntime,
    private val registry: NodeRegistry,
) : ApplicationRunner {
    // The runtime signals its observation publisher through a weak reference, so the
    // stream stops as soon as the publisher is collected. The console watches node
    // lifecycles for as long as it runs, so it keeps the publisher for that long.
    private var observation: Flow.Publisher<ZLinkObservedStatus<ZLinkMeshNodeSnapshot>>? = null
    private val expiry =
        Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "zoneworld-report-expiry").apply { isDaemon = true }
        }

    // --8<-- [start:doc-zw-observe-peers]
    override fun run(args: ApplicationArguments) {
        val publisher = runtime.observe(ZoneWorldNames.MESH, 32)
        observation = publisher
        publisher.subscribe(
            object : Flow.Subscriber<ZLinkObservedStatus<ZLinkMeshNodeSnapshot>> {
                override fun onSubscribe(subscription: Flow.Subscription) {
                    subscription.request(Long.MAX_VALUE)
                }

                override fun onNext(observed: ZLinkObservedStatus<ZLinkMeshNodeSnapshot>) {
                    registry.applyLiveRoutingIds(readyRoutingIds(observed.status()))
                }

                override fun onError(error: Throwable) {
                    println(
                        "node observation error mesh=${ZoneWorldNames.MESH}" +
                            " detail=${error.message}"
                    )
                }

                override fun onComplete() {}
            }
        )
        expiry.scheduleAtFixedRate(registry::expireStaleReports, 1, 1, TimeUnit.SECONDS)
    }

    // --8<-- [end:doc-zw-observe-peers]

    private fun readyRoutingIds(status: ZLinkMeshNodeSnapshot): Set<String> {
        val observed = linkedSetOf<String>()
        status
            .peers()
            .filter { it.state() == ZLinkPeerState.READY }
            .forEach { observed += it.nodeRid().toString() }
        return observed
    }
}
